core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -8482,11 +8497,23 @@ public void addSystemRole(String roleName, String[] userList, Permission[] permi
             return roleList;
         }
 
-        return getUserRoles(userName, filter);
+        String usernameWithTenantDomain = userName + "@" + this.getTenantDomain(this.tenantId);
+        synchronized (usernameWithTenantDomain.intern()) {
+            roleList = getRoleListOfUserFromCache(this.tenantId, userName);
```

Comment:
roleList = getRoleListOfUserFromCache(this.tenantId, userName); has to happen outside the synchronise block. Otherwise the synchronise block is called for each request and thus may cause unnecessary locks.